### ISSUE TITLE
Custom optimizer using closure

### ISSUE THREAD

**Issue by waddington-ou-phd-1** [open-mmlab/mmcv#866]
Hello,

I have implemented a custom optimizer in PyTorch that makes use of a closure function (https://pytorch.org/docs/stable/optim.html#optimizer-step-closure) to re-evaluate the model several times during a single optimizer `.step()` call.

I am now trying to move my implementation into MMDetection and I suspect the part that I need help with is to do with MMCV, hence I post here.

In the Optimizer hooks class, none of the calls to an optimizers `step()` function pass a closure function, so I suspect I need a new hook. 

I see there is a `ClosureHook` class, but it is not clear what this does, there is no documentation for it, and it is never used anywhere in code so there are no examples for me to work from. I'm not even sure how I would instantiate it or if it does what I think it would do.

My optimizer should be able to do the following inside a single call to its step() function:

1.  Change the weights of the model several times
2. Re-evaluate the model on the same input batch several times (forward pass and calculate loss)
3. Know the new loss value after changing the weights

I appreciate any guidance on this.
Thanks

**Comment by waddington-ou-phd-1** [comment]
I also leave this here to show that PyTorch Ignite can handle custom closures https://github.com/pytorch/ignite/issues/610

**Comment by MeowZheng** [comment]
Hi,

Many thanks for your good question. `ClosureHook` is not working for optimizer step function and we will add docstring to explain its usage.

For your requirements, my suggestions are the following.

- Overriding `after_train_iter` in optimizer hook and adding inner closure function.
- Overriding `BaseRunner`, as you need to reevaluate model several times before optimizer updating weights, the training pipelines in `EpochBasedRunner` and `IterBasedRunner` may not suit you.

Hope these are helpful for you. 

Best regards,
Meow
